packages/artillery-plugin-publish-metrics/lib/open-telemetry/index.js

```diff
@@ -0,0 +1,172 @@
+    this.events = events;
+    this.protocols = {
+      'http-proto': 'proto',
+      'http-json': 'http'
+    };
+
+    this.resource = Resource.default().merge(
+      new Resource({
+        [SemanticResourceAttributes.SERVICE_NAME]:
+          config.serviceName || 'Artillery-test'
+      })
+    );
+
+    if (config.traces) {
+      this.tracing = true;
+      this.traceConfig = config.traces;
+
+      this.configureTrace(this.traceConfig);
+
+      attachScenarioHooks(script, [
+        {
+          type: 'beforeRequest',
+          name: 'startOTelSpan',
+          hook: this.startOTelSpan.bind(this)
+        }
+      ]);
+
+      attachScenarioHooks(script, [
+        {
+          type: 'afterResponse',
+          name: 'exportOTelSpan',
+          hook: this.exportOTelSpan.bind(this)
+        }
+      ]);
+    }
+  }
+
+  configureTrace(config) {
+    debug('Configuring Tracer Provider');
+    const {
+      BasicTracerProvider,
+      TraceIdRatioBasedSampler,
+      ParentBasedSampler,
+      BatchSpanProcessor
+    } = require('@opentelemetry/sdk-trace-base');
+
+    this.tracerOpts = {
+      resource: this.resource
+    };
+    if (config.sampleRate) {
+      this.tracerOpts.sampler = new ParentBasedSampler({
+        root: new TraceIdRatioBasedSampler(config.sampleRate)
+      });
+    }
+
+    this.tracerProvider = new BasicTracerProvider(this.tracerOpts);
+
+    debug('Configuring Exporter');
+    this.traceExporterOpts = {};
+    if (config.endpoint) {
+      this.traceExporterOpts.url = config.endpoint;
+    }
+
+    if (config.headers) {
+      this.traceExporterOpts.headers = config.headers;
+    }
+
+    const protocol = !config.protocol
+      ? 'http'
+      : this.protocols[config.protocol];
+
+    const {
+      OTLPTraceExporter
+    } = require(`@opentelemetry/exporter-trace-otlp-${protocol}`);
+    this.exporter = new OTLPTraceExporter(this.traceExporterOpts);
+
+    this.tracerProvider.addSpanProcessor(
+      new BatchSpanProcessor(this.exporter, {
+        scheduledDelayMillis: 1000
+      })
+    );
+    this.tracerProvider.register();
+  }
+
+  startOTelSpan(req, userContext, events, done) {
+    debug('Starting span');
+    const startTime = Date.now();
+    userContext.vars['__otlStartTime'] = startTime;
+    const span = trace
+      .getTracer('artillery-tracer')
+      .startSpan('http_request', { startTime });
+    span.setAttribute('kind', 'client');
+    span.addEvent('http_request_started', startTime);
+    userContext.vars['__otlpSpan'] = span;
+
+    return done();
+  }
+
+  exportOTelSpan(req, res, userContext, events, done) {
+    if (!userContext.vars['__otlpSpan']) {
+      return done();
+    }
+
+    const span = userContext.vars['__otlpSpan'];
+    let endTime;
+
+    if (res.timings && res.timings.phases) {
+      span.setAttribute('responseTimeMs', res.timings.phases.firstByte);
+      endTime =
+        userContext.vars['__otlStartTime'] + res.timings.phases.firstByte;
```

Comment:
I'm choosing to calculate `endTime` differently in the otel reporter. In working on this, I noticed other tracing reporters (e.g. lightstep) occasionally had a big disparity between `responseTimeMs` values and the actual reported traces. (this may need to be fixed in other tracing reporters too)

I think this is because they were calling `span.finish()` in an afterHook, and so it could include javascript latency.
i.e. `beforeHook` starts span -> javascript runs including the request -> other hooks run -> `afterHook` finishes span

Now we set the start and end time of the span ourselves to match the request timings as much as possible.

Reply:
Great!